libmamba/include/mamba/core/download.hpp

```diff
@@ -82,29 +84,47 @@ namespace mamba
         );
     };
 
-    using DownloadRequestList = std::vector<DownloadRequest>;
-
-    struct MultiDownloadRequest
-    {
-        DownloadRequestList requests;
-    };
+    using MultiDownloadRequest = std::vector<DownloadRequest>;
 
     using DownloadResult = tl::expected<DownloadSuccess, DownloadError>;
-    using DownloadResultList = std::vector<DownloadResult>;
-
-    struct MultiDownloadResult
-    {
-        DownloadResultList results;
-    };
+    using MultiDownloadResult = std::vector<DownloadResult>;
 
     struct DownloadOptions
     {
+        using termination_function = std::optional<std::function<void()>>;
+
         bool fail_fast = false;
         bool sort = true;
+        termination_function on_unexpected_termination = std::nullopt;
+    };
+
+    class DownloadMonitor
+    {
+    public:
+
+        virtual ~DownloadMonitor() = default;
+
+        DownloadMonitor(const DownloadMonitor&) = delete;
+        DownloadMonitor& operator=(const DownloadMonitor&) = delete;
+        DownloadMonitor(DownloadMonitor&&) = delete;
+        DownloadMonitor& operator=(DownloadMonitor&&) = delete;
+
+        virtual void observe(MultiDownloadRequest& requests, DownloadOptions& options) = 0;
```

Comment:
Prefer private virtuality (calling the virtual function from a non-virtual function).